Fix Teleport ending trainer battles

AlexOn1ine opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Fix by redirecting to the Baton Pass script when it's a trainer battle.
## Images
<!-- Please provide with relevant GIFs or images to make it easier for reviewers to accept your PR quicker.-->
<!-- If it doesn't apply, feel free to remove this section. -->

## Issue(s) that this PR fixes
Fixes  #3164

## **Discord contact info**
<!--- formatted as name#numbers, e.g. Lunos#4026 -->
<!--- Contributors must join https://discord.gg/6CzjAG6GZk -->

AsparagusEduardo commented:
Redirecting to Baton Pass' script will reduce PP twice

LOuroboros commented:
You should update `CanTeleport` directly imo, @AlexOn1ine.

AsparagusEduardo commented:
> ```diff
>     switch (GetBattlerSide(battlerId))
>     {
>     case B_SIDE_OPPONENT:
> -       if (count == 1 || gBattleTypeFlags & BATTLE_TYPE_DOUBLE)
> +       if (count == 1 || gBattleTypeFlags & (BATTLE_TYPE_DOUBLE | BATTLE_TYPE_TRAINER))
>             return FALSE;
>         break;
> ```

This implies that Teleport fails in trainer battles for the opponent, which is not the case.

AlexOn1ine commented:
> Redirecting to Baton Pass' script will reduce PP twice

Fixed

> EDIT2: Also, I intended to do this at some point, but I guess I never ended up pushing it. canteleport and the subsequent jumpifbyte could be turned into a macro like jumpifcanteleport or something like that.

I think `canteleport` isn't needed anymore so I removed it completely.

LOuroboros commented:
> > Redirecting to Baton Pass' script will reduce PP twice
> 
> Fixed
> 
> > EDIT2: Also, I intended to do this at some point, but I guess I never ended up pushing it. canteleport and the subsequent jumpifbyte could be turned into a macro like jumpifcanteleport or something like that.
> 
> I think `canteleport` isn't needed anymore so I removed it completely.

Noice!

The suggestion can still be applied to `getbattlerside` which could be used for a `jumpifside` macro to save a few lines in the file though :eyes:

AlexOn1ine commented:
> The suggestion can still be applied to getbattlerside which could be used for a jumpifside macro to save a few lines in the file though eyes

Something like this? 
```diff
+	.macro jumpifside battler:req, side:req, jumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \jumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass
        jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
+       jumpifside BS_ATTACKER, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

LOuroboros commented:
> Something like this?

Yup. I guess that adds a pointless function call to `getbattlerside` though, which could be worked around like this:

```diff
+	.macro jumpifside battler:req, side:req, equalJumpInstr:req, notEqualJumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
        jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

AlexOn1ine commented:
> Yup. I guess that adds a pointless function call to getbattlerside though, which could be worked around like this:

Looks good, I like it but I think a small modification is needed to avoid the same issue the PR is trying to solve where Teleport would end a trainer battle.

Could `jumpifbattletype` moved inside the macro? So...
```diff
+	.macro jumpifside battler:req, side:req, equalJumpInstr:req, notEqualJumpInstr:req
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm

-       getbattlerside BS_ATTACKER
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+       jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
+       jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
-       jumpifbattletype BATTLE_TYPE_TRAINER, BattleScript_EffectBatonPass
-       jumpifbyte CMP_EQUAL, gBattleCommunication, B_SIDE_OPPONENT, BattleScript_EffectTeleportTryToRunAway
```

AlexOn1ine commented:
Added macro as suggested by Lunos.

AsparagusEduardo commented on asm/macros/battle_script.inc:
```diff
@@ -2177,6 +2171,12 @@
+	getbattlerside \battler
+	jumpifbyte CMP_EQUAL, gBattleCommunication, \side, \equalJumpInstr
+	jumpifbyte CMP_NOT_EQUAL, gBattleCommunication, \side, \notEqualJumpInstr
+	.endm
```
Uuh, aren't the jumpif macros meant to let the script continue to the next instruction if the condition doesn't pass?

LOuroboros replied:
They're meant to be used however the situation requires it. It's no different than doing a `call_if_eq` followed by a `call_if_ne` inside an overworld script.

In this case, the macro `jumpifside` first calls `getbattlerside` to write the side that the specified battler is on to `gBattleCommunication`, and then proceeds to jump to one battle script or another based on its value *(which is either `B_SIDE_PLAYER` or `B_SIDE_OPPONENT`)*.

AsparagusEduardo replied:
Yes, but the `goto BattleScript_ButItFailed` is never reached, so the second condition might as well be converted to:
```diff
-	jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass, BattleScript_EffectTeleportTryToRunAway
-	goto BattleScript_ButItFailed
+	jumpifside BS_ATTACKER, B_SIDE_PLAYER, BattleScript_EffectBatonPass
+	goto BattleScript_EffectTeleportTryToRunAway
```

AlexOn1ine replied:
Applied

AsparagusEduardo replied:
Wait a second, I just realized that `BattleScript_EffectTeleportTryToRunAway` is right below it, so you can just remove that goto ^^;
![image](https://github.com/rh-hideout/pokeemerald-expansion/assets/2904965/1226b8c4-f7e3-4b7c-93ac-3a21c1e3a548)
